remoting/rico-remoting-client/src/main/java/dev/rico/internal/remoting/client/legacy/communication/AbstractClientConnector.java

```diff
@@ -232,60 +226,53 @@ protected void release() {
         }
 
         releaseNeeded.set(false);// release is under way
-        backgroundExecutor.execute(new Runnable() {
-            @Override
-            public void run() {
+
+        backgroundExecutor.execute(() -> {
+            connectedFlagLock.lock();
+            if (connectedFlag.get()) {
                 try {
                     final List<Command> releaseCommandList = new ArrayList<>(Collections.singletonList(releaseCommand));
                     transmit(releaseCommandList);
                 } catch (RemotingException e) {
                     handleError(e);
+                } finally {
+                    connectedFlagLock.unlock();
                 }
             }
         });
     }
 
     public void connect(final boolean longPoll) {
-        if (connectedFlag.get()) {
-            throw new IllegalStateException("Can not call connect on a connected connection");
-        }
-
-        connectedFlag.set(true);
-        uiExecutor.execute(new Runnable() {
-            @Override
-            public void run() {
-                connectionFlagForUiExecutor = true;
+        connectedFlagLock.lock();
+        try {
+            if (connectedFlag.get()) {
+                throw new IllegalStateException("Can not call connect on a connected connection");
             }
-        });
 
-        backgroundExecutor.execute(new Runnable() {
-            @Override
-            public void run() {
-                commandProcessing();
-            }
-        });
-        useLongPolling.set(longPoll);
+            connectedFlag.set(true);
+            uiExecutor.execute(() -> connectionFlagForUiExecutor = true);
```

Comment:
Can this all be done in parallel? Maybe the tasks that are submitted to the uiExecutor and the backgroundExecutor will executed long after the lock is released

Reply:
Why is the change of the flag executed in the uiExecutor in the first place?
switching the flag should be possible from any thread